hi-ml-histopathology/src/histopathology/utils/report_utils.py

```diff
@@ -0,0 +1,339 @@
+    """
+    # Save metrics as a pickle because complex dataframe structure is lost in CSV
+    metrics_pickle = download_dir / "aml_metrics.pickle"
+    if metrics_pickle.is_file():
+        print(f"AML metrics file already exists at {metrics_pickle}")
+        with open(metrics_pickle, 'rb') as f:
+            metrics_df = pickle.load(f)
+    else:
+        metrics_df = aggregate_hyperdrive_metrics(run_id=parent_run_id,
+                                                  child_run_arg_name=crossval_arg_name,
+                                                  aml_workspace=aml_workspace)
+        metrics_pickle.parent.mkdir(parents=True, exist_ok=True)
+        print(f"Writing AML metrics file to {metrics_pickle}")
+        with open(metrics_pickle, 'wb') as f:
+            pickle.dump(metrics_df, f)
+    return metrics_df.sort_index(axis='columns')
+
+
+def plot_roc_curve(labels: Sequence, scores: Sequence, label: str, ax: Axes) -> None:
+    """Plot ROC curve for the given labels and scores, with AUROC in the line legend.
+
+    :param labels: The true binary labels.
+    :param scores: Scores predicted by the model.
+    :param label: An line identifier to be displayed in the legend.
+    :param ax: `Axes` object onto which to plot.
+    """
+    fpr, tpr, _ = roc_curve(labels, scores)
+    auroc = auc(fpr, tpr)
+    label = f"{label} (AUROC: {auroc:.3f})"
+    ax.plot(fpr, tpr, label=label)
+
+
+def plot_pr_curve(labels: Sequence, scores: Sequence, label: str, ax: Axes) -> None:
+    """Plot precision-recall curve for the given labels and scores, with AUROC in the line legend.
+
+    :param labels: The true binary labels.
+    :param scores: Scores predicted by the model.
+    :param label: An line identifier to be displayed in the legend.
+    :param ax: `Axes` object onto which to plot.
+    """
+    precision, recall, _ = precision_recall_curve(labels, scores)
+    aupr = auc(recall, precision)
+    label = f"{label} (AUPR: {aupr:.3f})"
+    ax.plot(recall, precision, label=label)
+
+
+def format_pr_or_roc_axes(plot_type: str, ax: Axes) -> None:
+    """Format PR or ROC plot with appropriate axis labels, limits, and grid.
+
+    :param plot_type: Either 'pr' or 'roc'.
+    :param ax: `Axes` object to format.
+    """
+    if plot_type == 'pr':
+        xlabel, ylabel = "Recall", "Precision"
+    elif plot_type == 'roc':
+        xlabel, ylabel = "False positive rate", "True positive rate"
+    else:
+        raise ValueError(f"Plot type must be either 'pr' or 'roc' (received '{plot_type}')")
+    ax.set_xlabel(xlabel)
+    ax.set_ylabel(ylabel)
+    ax.set_aspect(1)
+    ax.set_xlim(-.05, 1.05)
+    ax.set_ylim(-.05, 1.05)
+    ax.grid(color='0.9')
+
+
+def _plot_crossval_roc_and_pr_curves(crossval_dfs: Dict[int, pd.DataFrame],
+                                     roc_ax: Axes, pr_ax: Axes) -> None:
+    for k, tiles_df in crossval_dfs.items():
```

Comment:
docstring missing

Reply:
We are trying to be scrupulous with documentation, including private functions, for the purpose of helping future developers (/our future selves). It can be the same docstring copied and pasted from the wrapper function if that is appropriate?

Reply:
Sure, no problem